File: src/services/parsers/utils/OrderBlockTracker.ts
/**
 * Tracks order references (Vs. ORDINE) throughout PDF parsing
 *
 * CRITICAL BEHAVIOR: Order references persist across multiple invoice positions
 * until a new order block is encountered. This matches the PDF structure where
 * "Vs. ORDINE Nr. 10153" applies to all following positions until a new order appears.
 */
export class OrderBlockTracker {
  private currentOrders: string[] = [];

  /**
   * Start a new order block with given order references
   * @param orderReferences - Array of order numbers (e.g., ["10153", "10154"])
   */
  startNewBlock(orderReferences: string[]): void {
    this.currentOrders = orderReferences.filter(o => o.trim().length > 0);
  }

  /**
   * Get current order references for the current position
   * @returns Copy of current orders array (does NOT reset after retrieval)
   *
   * IMPORTANT: Returns a copy to prevent external modification.
   * Orders persist until startNewBlock() is called again.
   */
  getOrdersForPosition(): string[] {
    return [...this.currentOrders];
  }

  /**
   * Check if there are any active orders
   */
  hasOrders(): boolean {
    return this.currentOrders.length > 0;
  }

  /**
   * Reset tracker (typically at start of new PDF parsing)
   */
  reset(): void {
    this.currentOrders = [];
  }

  /**
   * Get current order count for debugging
   */
  getCurrentOrderCount(): number {
    return this.currentOrders.length;
  }
}

/**
 * Extract order references from a line containing "Vs. ORDINE"
 * @param line - Line containing order reference(s)
 * @returns Array of order numbers
 * @example
 * extractOrderReferences("Vs. ORDINE Nr. 10153") // Returns: ["10153"]
 * extractOrderReferences("Vs. ORDINE 0_10170_173_172") // Returns: ["10170", "173", "172"]
 */
export function extractOrderReferences(line: string): string[] {
  const candidates: string[] = [];

  // Underscore format, e.g. "0_10170_173_172" -> 10170, 10173, 10172
  const underscoreMatch = line.match(/(\d+(?:_\d+)+)/);
  if (underscoreMatch) {
    const parts = underscoreMatch[1].split('_');
    let basePrefix = '';
    for (const part of parts) {
      if (part.length === 5 && part.startsWith('10')) {
        candidates.push(part);
        basePrefix = part.slice(0, 2);
      } else if (part.length === 3 && basePrefix) {
        candidates.push(basePrefix + part);
      }
    }
  }

  // Standalone 10xxx numbers
  const standaloneMatches = line.match(/\b(10\d{3})\b/g) ?? [];
  for (const order of standaloneMatches) {
    if (!candidates.includes(order)) {
      candidates.push(order);
    }
  }

  return candidates;
}

File: logicdev_PDF-Parser_V1/logicdev_Core/order_block_tracker.py
"""
Order block tracking for invoice parsing.

Port of OrderBlockTracker.ts.
CRITICAL: Order numbers persist across positions until a new "Vs. ORDINE" appears.
"""
import re
from dataclasses import dataclass, field
from typing import Literal


OrderStatus = Literal["YES", "NO", "check"]

ORDER_REFERENCE_PATTERN = re.compile(r'Vs\.\s*ORDINE', re.IGNORECASE)


@dataclass
class OrderBlockResult:
    """Result of order lookup for a position."""
    primary: str | None = None
    raw: list[str] = field(default_factory=list)


def extract_order_candidates(text: str) -> list[str]:
    """
    Extract order candidates from order reference lines.

    Handles:
    - "Vs. ORDINE Nr. 10153"
    - "Vs. ORDINE 0_10170_173_172" (underscore-separated with short codes)
    - Multiple direct 10xxx numbers
    """
    candidates: list[str] = []

    # Look for underscore-separated format: 0_10170_173_172
    underscore_match = re.search(r'(\d+(?:_\d+)+)', text)
    if underscore_match:
        parts = underscore_match.group(1).split("_")
        base_prefix = ""
        for part in parts:
            if len(part) == 5 and part.startswith("10"):
                candidates.append(part)
                base_prefix = part[:2]
            elif len(part) == 3 and base_prefix:
                # Short code like 173 becomes 10173
                candidates.append(base_prefix + part)

    # Also extract standalone 10xxx numbers
    for match in re.finditer(r'\b(10\d{3})\b', text):
        if match.group(1) not in candidates:
            candidates.append(match.group(1))

    return candidates


def get_order_status(candidates: list[str]) -> OrderStatus:
    """Get order status based on candidates count."""
    if len(candidates) == 0:
        return "NO"
    if len(candidates) == 1:
        return "YES"
    return "check"


class OrderBlockTracker:
    """
    Tracks order numbers across position blocks.

    CRITICAL: Order numbers are NOT reset after each position!
    They persist until a new "Vs. ORDINE" line appears.
    """

    def __init__(self):
        self._current_block_orders: list[str] = []
        self._order_history: list[dict] = []
        self._position_count: int = 0

    def start_new_block(self, order_numbers: list[str]) -> None:
        """Start a new order block when 'Vs. ORDINE' is found."""
        if order_numbers:
            self._current_block_orders = list(order_numbers)
            self._order_history.append({
                "orders": list(order_numbers),
                "start_position": self._position_count + 1,
            })

    def get_orders_for_position(self) -> OrderBlockResult:
        """
        Get orders for the current position.
        IMPORTANT: Does NOT reset the order block!
        """
        self._position_count += 1
        return OrderBlockResult(
            primary=self._current_block_orders[-1] if self._current_block_orders else None,
            raw=list(self._current_block_orders),
        )

    @property
    def position_count(self) -> int:
        return self._position_count

    @property
    def order_history(self) -> list[dict]:
        return list(self._order_history)

    @property
    def has_active_orders(self) -> bool:
        return len(self._current_block_orders) > 0

    def reset(self) -> None:
        """Reset tracker for new parsing session."""
        self._current_block_orders = []
        self._order_history = []
        self._position_count = 0
